src/Chainweb/Pact/Backend/PactState/GrandHash.hs

```diff
@@ -0,0 +1,753 @@
+            else do
+              ensureBlockHeightExists db target
+              atomicModifyIORef' targetsRef $ \m -> (Map.insertWith Set.union cid (Set.singleton target) m, ())
+
+      readIORef targetsRef
+
+-- | Resolve a single target.
+--
+--   Returns the resolved BlockHeight, along with the Set of ChainIds that
+--   contain that target.
+resolveTarget :: (Logger logger)
+  => logger
+  -> [ChainId]
+  -> FilePath
+  -> TargetBlockHeight
+  -> IO (Set ChainId, BlockHeight)
+resolveTarget logger cids pactDir target = do
+  let targets = case target of
+        Latest -> LatestAll
+        Target t -> TargetAll (Set.singleton t)
+  resolved <- resolveTargets logger cids pactDir targets
+
+  when (Map.size resolved == 0) $ do
+    error "internal logic error in resolveTarget: resolved was empty"
+
+  when (not $ allEqual $ Map.elems resolved) $ do
+    error "internal logic error in resolveTarget: targets are not the same across all chains"
+
+  -- kinda gross.
+  let height = List.head $ Set.toList $ List.head $ Map.elems resolved
+
+  pure (Set.fromList (Map.keys resolved), height)
+  where
+    allEqual :: Eq a => [a] -> Bool
+    allEqual = and . mapAdjacent (==)
+
+    mapAdjacent :: (a -> a -> b) -> [a] -> [b]
+    mapAdjacent f xs = zipWith f xs (tail xs)
+
+-- | Compute the GrandHashes at the specified targets.
+--
+--   The targets can either be the latest BlockHeight only,
+--   or a set of specified targets (for computing many grand hashes per chain at
+--   once).
+--
+--   The list this returns is in descending order by the BlockHeight.
+--   This is to facilitate easier searching in the style of 'findGrandHash'.
+computeGrandHashesAt :: (Logger logger)
+  => logger
+  -> [ChainId]
+  -> FilePath
+  -> FilePath
+  -> BlockHeightTargets
+  -> ChainwebVersion
+  -> IO [(BlockHeight, Map ChainId (ByteString, BlockHeader))]
+computeGrandHashesAt logger cids pactDir rocksDir ts chainwebVersion = do
+
+  -- We do this in three phases;
+  --   1) In the first phase, we check that the db matches
+  --      our expectations (blockheights we intend to use are
+  --      available, etc) and resolve targets across all chains.
+  --   2) In the second phase, we gather all of the grand hashes.
+  --   2) In the third phase, we pair all of the grand hashes with the associated
+  --      block headers.
+
+  -- Phase 1:
+  --   Check that the DB matches our expectations and resolve
+  --   any slight mismatches between chains
+  chainTargets <- resolveTargets logger cids pactDir ts
+
+  -- Phase 2:
+  --   Gather all of the grand hashes
+
+  -- This holds the Grand Hashes for each chain, at each BlockHeight of
+  -- interest.
+  chainHashesRef <- newIORef @(Map BlockHeight (Map ChainId ByteString)) Map.empty
+
+  -- It's simpler to instantiate the keys ahead of time
+  --forM_ targets $ \target -> do
+  --  modifyIORef' chainHashesRef $ \m -> Map.insert target Map.empty m
+
```

Comment:
```suggestion

```

Not only is instantiating the keys ahead of time incorrect (because some of them may not exist), but this comment is just redundant.